ambry-account/src/main/java/com/github/ambry/account/mysql/MySqlDataAccessor.java

```diff
@@ -54,49 +61,91 @@
+    sortedDbEndpoints = inputEndpoints;
+    if (!sortedDbEndpoints.get(0).isWriteable()) {
+      throw new IllegalArgumentException("No endpoints are writable");
+    }
+
+    initializeDriver(sortedDbEndpoints.get(0).getUrl());
+
     // AccountService needs to work if mysql is down.  Mysql can also reboot.
     try {
-      getDatabaseConnection();
+      getDatabaseConnection(true);
     } catch (SQLException e) {
       if (isCredentialError(e)) {
         throw e;
       } else {
-        // try again later
+        logger.error("No writable database available, will retry later.");
       }
     }
   }
 
-  /** Test constructor */
-  public MySqlDataAccessor(MySqlUtils.DbEndpoint dbEndpoint, Driver mysqlDriver, MySqlAccountStoreMetrics metrics) {
-    mysqlUrl = dbEndpoint.getUrl();
-    mysqlUser = dbEndpoint.getUsername();
-    mysqlPassword = dbEndpoint.getPassword();
-    this.mysqlDriver = mysqlDriver;
-    this.metrics = metrics;
+  private void initializeDriver(String url) throws SQLException {
+    if (mysqlDriver == null) {
+      mysqlDriver = DriverManager.getDriver(url);
+    }
   }
 
   /**
-   * @return a JDBC {@link Connection} to the database.  An existing connection will be reused.
+   * @return a JDBC {@link Connection} to the database.  An existing connection will be reused,
+   * unless a connection to a better-ranked enpoint is available.
+   * @param needWritable whether the database instance needs to be writeable.
    * @throws SQLException
    */
-  public synchronized Connection getDatabaseConnection() throws SQLException {
-    if (activeConnection != null && activeConnection.isValid(5)) {
+  public synchronized Connection getDatabaseConnection(boolean needWritable) throws SQLException {
+
+    // Close active connection if no longer valid
+    if (activeConnection != null && !activeConnection.isValid(5)) {
+      reset();
+      activeConnection = null;
+      connectedEndpoint = null;
+    }
+
+    // If the active connection is good and it's the best endpoint, keep it.
+    if (activeConnection != null && !isBetterEndpoint(sortedDbEndpoints.get(0), connectedEndpoint)) {
       return activeConnection;
     }
-    if (activeConnection != null) {
-      activeConnection.close();
+    // See if we can do better
+    Pair<DbEndpoint, Connection> endpointConnectionPair = connectToBestAvailableEndpoint(needWritable);
+    if (connectedEndpoint == endpointConnectionPair.getFirst()) {
+      // No better endpoint  was available.
+      logger.debug("Still connected to {}", connectedEndpoint.getUrl());
+    } else {
+      // New connection established!
+      // TODO: mysqlMetrics.connectionFixedCount.inc();
+      connectedEndpoint = endpointConnectionPair.getFirst();
+      String qualifier = connectedEndpoint.isWriteable() ? "writable" : "read-only";
+      logger.info("Connected to {} enpoint: {}", qualifier, connectedEndpoint.getUrl());
+      closeQuietly(activeConnection); // TODO: reset?
```

Comment:
Yeah, would reset() be needed to clear PreparedStatements in 'statementCache' associated with this connection being closed?